smart-contracts/contracts/CosmosBridge.sol

```diff
@@ -107,47 +110,53 @@ contract CosmosBridge is CosmosBridgeStorage, Oracle {
                     amount,
                     doublePeg,
                     nonce,
-                    _networkDescriptor
+                    _networkDescriptor,
+                    tokenName,
+                    tokenSymbol,
+                    tokenDecimals
                 )
             )
         );
     }
-
+    
     function verifySignature(
         address signer,
         bytes32 hashDigest,
         uint8 _v,
-		bytes32 _r,
-		bytes32 _s
+		    bytes32 _r,
+		    bytes32 _s
     ) private pure returns (bool) {
-		bytes32 messageDigest = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hashDigest));
-		return signer == ecrecover(messageDigest, _v, _r, _s);
+		    bytes32 messageDigest = keccak256(abi.encodePacked("\x19Ethereum Signed Message:\n32", hashDigest));
+		    return signer == ecrecover(messageDigest, _v, _r, _s);
 	}
     
-    // this is unfortunately the best we can do to ensure no duplicate validators are calling
-    // it is possible to build a hashmap in memory, but I'm unsure of how much that saves and
-    // it would require some pretty low level work for this very simple function
-    // Alternatively, cast addresses to UINT's and possibly do some bitwise operations
-    // to ensure there are no duplicate numbers
-    function findDup(SignatureData[] calldata _validators) public pure returns (bool) {
+    function getSignedPowerAndFindDup(
+        SignatureData[] calldata _validators,
+        bytes32 hashDigest
+    ) private view returns(bool, uint256 pow) {
```

Comment:
Let's remove the boolean being returned here and instead just revert if a false would have been returned.